### PR TITLE
Add presentation addon

### PR DESCRIPTION
**Pull request by PhILoXs**
Nearly finished. But works as intended so far.

### CODE REVIEW

File: addons/presentation/XEH_postInit.sqf [hunk 1]
```diff
@@ -0,0 +1,133 @@
+#include "script_component.hpp"
+
+GVAR(topics) = [];
+//get all topics from config
+private _topics = configfile >> QGVAR(topics);
+
+private ["_topic", "_displayName", "_path", "_pages", "_extension"];
+for "_i" from 0 to (count _topics - 1) do {
```
Comment (Timi007) [inline addons/presentation/XEH_postInit.sqf]: Is there a reason for not using forEach?

File: addons/presentation/XEH_postInit.sqf [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+            "true",
+            2
+            ];
+    } forEach GVAR(controls);
+};
+
+
+
+[QGVAR(startVideo), {
+	params [["_video", "", [""]]];
```
Comment (Timi007) [inline addons/presentation/XEH_postInit.sqf:96]: 2 spaced tab.
Editorconfig not active?

File: addons/presentation/functions/fnc_changeSlide.sqf [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+ *  Parameter(s):
+ *      0: [TYPE] - [argument name]
+ *
+ *  Returns:
+ *      [TYPE] - [return name]
+ *
+ *  Example:
+ *      [[arguments]] call [function name]
+ *
+ */
```
Comment (Timi007) [inline addons/presentation/functions/fnc_changeSlide.sqf]: Missing documentation.

File: addons/presentation/functions/fnc_fillPagesCombo.sqf [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+ *
+ *  Parameter(s):
+ *      0: CONTROL - Listbox control
+ *      1: NUMBER - Index of selected entry
+ *
+ *  Returns:
+ *      Nothing.
+ *
+ *  Example:
+ *      [[arguments]] call [function name]
```
Comment (Timi007) [inline addons/presentation/functions/fnc_fillPagesCombo.sqf]: Missing example.

File: addons/presentation/functions/fnc_openDialog.sqf [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+ *      Opens the presentation selection dialog and add available topics.
+ *
+ *  Parameter(s):
+ *      None.
+ *
+ *  Returns:
+ *      Nothing.
+ *
+ *  Example:
+ *      [[arguments]] call [function name]
```
Comment (Timi007) [inline addons/presentation/functions/fnc_openDialog.sqf]: Missing example.

File: addons/presentation/functions/fnc_setDisplay.sqf [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+ *  Parameter(s):
+ *      0: [TYPE] - [argument name]
+ *
+ *  Returns:
+ *      [TYPE] - [return name]
+ *
+ *  Example:
+ *      [[arguments]] call [function name]
+ *
+ */
```
Comment (Timi007) [inline addons/presentation/functions/fnc_setDisplay.sqf]: Missing documentation.

File: addons/presentation/functions/fnc_setDisplay.sqf [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+        _displayData pushBack [_previewLblCtrl, [_x, _y]];
+        _displayData pushBack [_previewBgLblCtrl, [_x, _y]];
+    };
+};
+
+TRACE_1("", _displayData);
+{
+    (_x select 0) ctrlSetPosition (_x select 1);
+    (_x select 0) ctrlCommit _commitSpeed;
+} forEach _displayData;
```
Comment (Timi007) [inline addons/presentation/functions/fnc_setDisplay.sqf]: Add `params ["_ctrl", "_pos"]` for more readability.

File: addons/presentation/functions/fnc_setSlide.sqf [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+ *      Sets the given slide on all screens.
+ *
+ *  Parameter(s):
+ *      0: [TYPE] - [argument name]
+ *
+ *  Returns:
+ *      [TYPE] - [return name]
+ *
+ *  Example:
+ *      [[arguments]] call [function name]
```
Comment (Timi007) [inline addons/presentation/functions/fnc_setSlide.sqf]: Missing documentation.

File: addons/presentation/functions/fnc_setSlide.sqf [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+ *
+ */
+params [["_slide", "", [""]], ["_previewSlide", "", [""]]];
+TRACE_2("", _slide, _previewSlide);
+
+private _extension = _slide select [(count _slide) - 3 , 3];
+CHECK(!(_extension in [ARR_4("ogv", "jpg", "jpeg", "paa")]));
```
Comment (Timi007) [inline addons/presentation/functions/fnc_setSlide.sqf]:
```suggestion
CHECKRET(!(_extension in [ARR_4("ogv", "jpg", "jpeg", "paa")]), ERROR("Not supported media format"));
```

File: addons/presentation/functions/fnc_setSlide.sqf [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+
+//set world screens
+if 	(_extension isEqualTo "ogv") then {
+    //stop any video that could currently running
+    QGVAR(stopVideo) call CBA_fnc_globalEvent;
+    [QGVAR(startVideo), _slide] call CBA_fnc_globalEvent;
+};
+if (_extension in ["jpg", "jpeg", "paa"]) then {
```
Comment (Timi007) [inline addons/presentation/functions/fnc_setSlide.sqf]:
```suggestion
} else {
```
Remove `};` in line above.
No need for checking the extension again.

File: addons/presentation/functions/fnc_setSlide.sqf [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+private _extension = _slide select [(count _slide) - 3 , 3];
+CHECK(!(_extension in [ARR_4("ogv", "jpg", "jpeg", "paa")]));
+
+//set world screens
+if 	(_extension isEqualTo "ogv") then {
+    //stop any video that could currently running
+    QGVAR(stopVideo) call CBA_fnc_globalEvent;
+    [QGVAR(startVideo), _slide] call CBA_fnc_globalEvent;
+};
```
Comment (Timi007) [inline addons/presentation/functions/fnc_setSlide.sqf]:
```suggestion

```

File: addons/presentation/functions/fnc_setSlide.sqf [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+if (_extension in ["jpg", "jpeg", "paa"]) then {
+    {
+        _x setObjectTextureGlobal [0, _slide];
+        } forEach GVAR(screens);
+};
```
Comment (Timi007) [inline addons/presentation/functions/fnc_setSlide.sqf]: Indentation

File: addons/presentation/functions/fnc_startPresentation.sqf [hunk 13]
```diff
@@ -0,0 +1,108 @@
...
+ *      Starts the presentation.
+ *
+ *  Parameter(s):
+ *      0: [TYPE] - [argument name]
+ *
+ *  Returns:
+ *      [TYPE] - [return name]
+ *
+ *  Example:
+ *      [[arguments]] call [function name]
```
Comment (Timi007) [inline addons/presentation/functions/fnc_startPresentation.sqf]: Missing documentation.

File: addons/presentation/functions/fnc_stopPresentation.sqf [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+ *      Stops the presentation.
+ *
+ *  Parameter(s):
+ *      0: [TYPE] - [argument name]
+ *
+ *  Returns:
+ *      [TYPE] - [return name]
+ *
+ *  Example:
+ *      [[arguments]] call [function name]
```
Comment (Timi007) [inline addons/presentation/functions/fnc_stopPresentation.sqf]: Missing documentation.

File: addons/presentation/functions/fnc_stopPresentation.sqf [hunk 15]
```diff
@@ -0,0 +1,41 @@
...
+ *
+ */
+
+ params [["_disconnect", false , [false]]];
```
Comment (Timi007) [inline addons/presentation/functions/fnc_stopPresentation.sqf]:
```suggestion
 params [["_disconnect", false, [false]]];
```

File: addons/presentation/script_component.hpp [hunk 16]
```diff
@@ -0,0 +1,26 @@
+#define COMPONENT presentation
+#define COMPONENT_BEAUTIFIED Presentation
+#include "\z\mts_presentation\addons\presentation\script_mod.hpp"
+
+ #define DEBUG_MODE_FULL
```
Comment (Timi007) [inline addons/presentation/script_component.hpp]:
```suggestion
// #define DEBUG_MODE_FULL
```

File: addons/presentation/script_component.hpp [hunk 17]
```diff
@@ -0,0 +1,26 @@
...
+ #define DISABLE_COMPILE_CACHE
```
Comment (Timi007) [inline addons/presentation/script_component.hpp]:
```suggestion
// #define DISABLE_COMPILE_CACHE
```

File: addons/presentation/script_version.hpp [hunk 18]
```diff
@@ -0,0 +1,4 @@
+#define MAJOR 1
+#define MINOR 0
+#define PATCHLVL 0
+#define BUILD 190306
```
Comment (Timi007) [inline addons/presentation/script_version.hpp]: Don't forget to update build.

File: addons/presentation/ui/mts_presentation_dialog.hpp [hunk 19]
```diff
@@ -0,0 +1,147 @@
...
+        class ComboScrollBar;
+
+        colorBackground[] = {0.231,0.231,0.231,1};
+        colorPictureRight[] = {0, 0, 0, 1};
+        colorTextRight[] = {0, 0, 0, 1};
+        arrowEmpty = "\A3\ui_f\data\GUI\Cfg\Scrollbar\arrowEmpty_ca.paa";
+    	arrowFull = "\A3\ui_f\data\GUI\Cfg\Scrollbar\arrowFull_ca.paa";
+    	thumb = "\A3\ui_f\data\GUI\Cfg\Scrollbar\thumb_ca.paa";
+    	border = "\A3\ui_f\data\GUI\Cfg\Scrollbar\border_ca.paa";
+    	color[] = {1,1,1,0.6};
```
Comment (Timi007) [inline addons/presentation/ui/mts_presentation_dialog.hpp]: Indentation

File: addons/presentation/ui/mts_presentation_display.hpp [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+				idc = IDC_DISPLAY_PREVIEW_BACKGROUND_LABEL;
+				w = "(512 / 1920) * SafeZoneW";
+				h = "(40 / 1080) * SafeZoneH";
+			};
+			class preview_label: RscText {
+				idc = IDC_DISPLAY_PREVIEW_LABEL;
+				text = CSTRING(preview);
+				w = "(512 / 1920) * SafeZoneW";
+				h = "(40 / 1080) * SafeZoneH";
+			};
```
Comment (Timi007) [inline addons/presentation/ui/mts_presentation_display.hpp]: Indentation

**Review by Timi007 — CHANGES_REQUESTED** [review]
A lot of weird indentations for some files. Is editorconfig working?

German translation missing.

File: addons/presentation/XEH_postInit.sqf (hunk 1 above)
Comment (PhILoXs) [reply addons/presentation/XEH_postInit.sqf]: Afaik, forEach doesn't work for config entries.

File: addons/presentation/XEH_postInit.sqf (hunk 2 above)
Comment (PhILoXs) [reply addons/presentation/XEH_postInit.sqf:96]: > 2 spaced tab.
> Editorconfig not active?

Don't know why, but in the changes view it seems to be 2 spaces. In the file view of Github and in my editor it is 4 spaces.

File: addons/presentation/ui/mts_presentation_dialog.hpp (hunk 19 above)
Comment (PhILoXs) [reply addons/presentation/ui/mts_presentation_dialog.hpp]: Again, not on view file from Github and in my editor.

File: addons/presentation/ui/mts_presentation_display.hpp (hunk 20 above)
Comment (PhILoXs) [reply addons/presentation/ui/mts_presentation_display.hpp]: Again, not on view file from Github and in my editor.

File: addons/presentation/functions/fnc_getTopics.sqf [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+params ["_topics"];
+
+private ["_topic", "_displayName", "_path", "_name", "_pages", "_extension"];
+
+for "_i" from 0 to (count _topics - 1) do {
+     _topic = configName (_topics select _i);
+     _displayName = getText (_topics >> _topic >> "displayName");
+     _path = getText (_topics >>  _topic >> "path");
+     _name = getText (_topics >>  _topic >> "name");
+     _pages = getNumber (_topics >>  _topic >> "pages");
```
Comment (Timi007) [inline addons/presentation/functions/fnc_getTopics.sqf]:
```suggestion
     _pages = getNumber (_topics >> _topic >> "slides");
```

File: addons/presentation/functions/fnc_getTopics.sqf [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+ */
+
+params ["_topics"];
+
+private ["_topic", "_displayName", "_path", "_name", "_pages", "_extension"];
+
+for "_i" from 0 to (count _topics - 1) do {
+     _topic = configName (_topics select _i);
+     _displayName = getText (_topics >> _topic >> "displayName");
+     _path = getText (_topics >>  _topic >> "path");
```
Comment (Timi007) [inline addons/presentation/functions/fnc_getTopics.sqf]:
```suggestion
     _path = getText (_topics >> _topic >> "path");
```

File: addons/presentation/functions/fnc_getTopics.sqf [hunk 23]
```diff
@@ -0,0 +1,32 @@
...
+     _name = getText (_topics >>  _topic >> "name");
```
Comment (Timi007) [inline addons/presentation/functions/fnc_getTopics.sqf]:
```suggestion
     _name = getText (_topics >> _topic >> "name");
```

File: addons/presentation/functions/fnc_getTopics.sqf [hunk 24]
```diff
@@ -0,0 +1,32 @@
...
+     _pages = getNumber (_topics >>  _topic >> "pages");
+     _extension = toLower (getText (_topics >>  _topic >> "extension"));
```
Comment (Timi007) [inline addons/presentation/functions/fnc_getTopics.sqf]:
```suggestion
     _extension = toLower (getText (_topics >> _topic >> "extension"));
```

File: addons/presentation/functions/fnc_setDisplay.sqf [hunk 25]
```diff
@@ -0,0 +1,122 @@
...
+ *      Creates or changes size of displays for the presenter.
+ *
+ *  Parameter(s):
+ *      0: STRING - Displaysize. Current values: "small" or "big"
+ *
+ *  Returns:
+ *      Nothing.
+ *
+ *  Example:
+ *      "small" call mts_presentation_fnc_setDisplay
```
Comment (Timi007) [inline addons/presentation/functions/fnc_setDisplay.sqf]:
```suggestion
 *      ["small"] call mts_presentation_fnc_setDisplay
```

File: addons/presentation/functions/fnc_setSlide.sqf [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+ *
+ *  Parameter(s):
+ *      0: STRING - Full slide path. (e.g. \z\mts_slides\addons\slides\test_images\slide1.jpg)
+ *      1: STRING - Full slide path for the next slide (preview slide for the presenter) (e.g. \z\mts_slides\addons\slides\test_images\slide2.jpg)
+ *
+ *  Returns:
+ *      Nothing.
+ *
+ *  Example:
+ *      [\z\mts_slides\addons\slides\test_images\slide1.jpg, \z\mts_slides\addons\slides\test_images\slide2.jpg] call mts_presentation_fnc_setSlide
```
Comment (Timi007) [inline addons/presentation/functions/fnc_setSlide.sqf]:
```suggestion
 *      ["my_slide_mod\test_images\slide1.jpg", "my_slide_mod\test_images\slide2.jpg"] call mts_presentation_fnc_setSlide
```

File: addons/presentation/functions/fnc_setSlide.sqf [hunk 27]
```diff
@@ -0,0 +1,51 @@
+#include "script_component.hpp"
+/**
+ *  Author: PhILoX
+ *
+ *  Description:
+ *      Sets the given slide on all screens.
+ *
+ *  Parameter(s):
+ *      0: STRING - Full slide path. (e.g. \z\mts_slides\addons\slides\test_images\slide1.jpg)
```
Comment (Timi007) [inline addons/presentation/functions/fnc_setSlide.sqf]:
```suggestion
 *      0: STRING - Full slide path. (e.g. "my_slide_mod\test_images\slide1.jpg")
```

File: addons/presentation/functions/fnc_setSlide.sqf [hunk 28]
```diff
@@ -0,0 +1,51 @@
...
+ *      1: STRING - Full slide path for the next slide (preview slide for the presenter) (e.g. \z\mts_slides\addons\slides\test_images\slide2.jpg)
```
Comment (Timi007) [inline addons/presentation/functions/fnc_setSlide.sqf]:
```suggestion
 *      1: STRING - Full slide path for the next slide (preview slide for the presenter) (e.g. "my_slide_mod\test_images\slide2.jpg")
```

File: addons/presentation/functions/fnc_setSlide.sqf [hunk 29]
```diff
@@ -0,0 +1,51 @@
...
+ */
+params [["_slide", "", [""]], ["_previewSlide", "", [""]]];
+TRACE_2("", _slide, _previewSlide);
+
+private _extension = _slide select [(count _slide) - 3 , 3];
+CHECKRET(!(_extension in [ARR_3("ogv", "jpg", "paa")]), ERROR(format [ARR_2("Not supported media format: %1", _extension)]));
+
+
+//set world screens
+if 	(_extension isEqualTo "ogv") then {
```
Comment (Timi007) [inline addons/presentation/functions/fnc_setSlide.sqf]:
```suggestion
if (_extension isEqualTo "ogv") then {
```

File: addons/presentation/functions/fnc_setSlide.sqf [hunk 30]
```diff
@@ -0,0 +1,51 @@
...
+    //stop any video that could currently running
+    QGVAR(stopVideo) call CBA_fnc_globalEvent;
+    [QGVAR(startVideo), _slide] call CBA_fnc_globalEvent;
+} else {
+    {
+        _x setObjectTextureGlobal [0, _slide];
+    } forEach GVAR(screens);
+};
+
+//check locality
```
Comment (Timi007) [inline addons/presentation/functions/fnc_setSlide.sqf]: No locality is checked here. Remove comment?

File: addons/presentation/functions/fnc_startPresentation.sqf [hunk 31]
```diff
@@ -0,0 +1,111 @@
...
+private _presenter = GVAR(presenter);
+private _player = CURRENT_PLAYER;
+
+//check whether the controls are already occupied or the presenter is unable to present.
+if (!isNull _presenter) then {
+    if (!alive _presenter || (_presenter getVariable ["ACE_isUnconscious", false])) then {
+        _presenter = _player;
+    };
+    if !(_presenter isEqualTo _player) then {
+        [["Metis Presentation:", 1.2, [1, 1, 0, 1]], LLSTRING(presenter_slot_occupied_line_1), format[LLSTRING(presenter_slot_occupied_line_2), name _presenter], false] call CBA_fnc_notify;
```
Comment (Timi007) [inline addons/presentation/functions/fnc_startPresentation.sqf]: Move "Metis Presentation" to stringtable

File: addons/presentation/functions/fnc_startPresentation.sqf [hunk 32]
```diff
@@ -0,0 +1,111 @@
...
+            breakOut "function";
+        };
+    };
+    [true] call FUNC(stopPresentation);
+
+    missionNamespace setVariable [QGVAR(topic), _topic, true];
+    missionNamespace setVariable [QGVAR(currentPage), _page, true];
+
+    if (GVAR(screens) isEqualTo []) then {
+        [["Metis Presentation:", 1.2, [1, 1, 0, 1]], LLSTRING(no_screen_to_display), false] call CBA_fnc_notify;
```
Comment (Timi007) [inline addons/presentation/functions/fnc_startPresentation.sqf]: Same as above

File: addons/presentation/functions/fnc_stopPresentation.sqf [hunk 33]
```diff
@@ -0,0 +1,42 @@
...
+ *      Nothing.
+ *
+ *  Example:
+ *      call mts_presentation_fnc_stopPresentation
+ *
+ */
+
+ params [["_notLocal", false, [false]]];
+
+if (!_notLocal) then {
```
Comment (Timi007) [inline addons/presentation/functions/fnc_stopPresentation.sqf]: Reverse if for less nesting -> if (_notLocal) exitWith {};

File: addons/presentation/initSettings.hpp [hunk 34]
```diff
@@ -0,0 +1,84 @@
+//params ["_setting", "_value"];
```
Comment (Timi007) [inline addons/presentation/initSettings.hpp]: Remove comment

File: addons/presentation/initSettings.hpp [hunk 35]
```diff
@@ -0,0 +1,84 @@
...
+//edit change setting on the fly
+// [
+//     QGVAR(showTimer),
+//     "CHECKBOX",
+//     [LLSTRING(cba_settings_show_timer), LLSTRING(cba_settings_show_timer_tooltip)],
+//     LLSTRING(cba_category_name),
+//     true,
+//     0,
+//     {}
+// ] call CBA_settings_fnc_init;
```
Comment (Timi007) [inline addons/presentation/initSettings.hpp]: Either remove setting or implement it.

File: addons/presentation/ui/mts_presentation_dialog.hpp [hunk 36]
```diff
@@ -0,0 +1,147 @@
...
+		};
+		class listbox: Listbox {
+			idc = IDC_LISTBOX;
+            style = "16";
+			x = "SafeZoneX + (736 / 1920) * SafeZoneW";
+			y = "SafeZoneY + (475 / 1080) * SafeZoneH";
+			w = "(264 / 1920) * SafeZoneW";
+			h = "(192 / 1080) * SafeZoneH";
+		};
+    };
```
Comment (Timi007) [inline addons/presentation/ui/mts_presentation_dialog.hpp]: Inconsistent indentation. Maybe mixed tabs + spaces? Try converting all to spaces.

File: addons/presentation/ui/mts_presentation_display.hpp [hunk 37]
```diff
@@ -0,0 +1,45 @@
...
+			class preview_label: RscText {
+				idc = IDC_DISPLAY_PREVIEW_LABEL;
+				text = CSTRING(display_preview);
+				w = "(512 / 1920) * SafeZoneW";
+				h = "(40 / 1080) * SafeZoneH";
+			};
+		};
+	};
```
Comment (Timi007) [inline addons/presentation/ui/mts_presentation_display.hpp]: Again inconsitent indentation.

**Review by Timi007 — CHANGES_REQUESTED** [review]
Some findings. Also missing german translation

**Review by Timi007 — APPROVED** [review]

**Pull request merged**